src/main.rs

```diff
@@ -1,3 +1,196 @@
+        let mut buf = [0; 8];
+        self.stream.read_exact(&mut buf)?;
+        let id = u64::from_le_bytes(buf);
+        println!("Message ID: {}", id);
+
+        let mut buf = [0; 1];
+        self.stream.read_exact(&mut buf)?;
+        let typ = RespType::from_u8(buf[0]).expect("Invalid response type!");
+        println!("Message type: {:?}", typ);
+        match typ {
+            RespType::RespOk => (),
+            RespType::RespErr => {
+                let mut buf = [0; 4];
+                self.stream.read_exact(&mut buf)?;
+                let code = u32::from_le_bytes(buf);
+                println!("Error code: {}", code);
+            }
+            RespType::NotifyOutputAvailable => (),
+            RespType::NotifyProcessDied => (),
+        };
+        Ok(())
+    }
+
+    fn quit(&mut self) -> Result<(), io::Error> {
+        let mut buf = GuestAgent::create_header(self.get_new_msg_id(), MsgType::MsgQuit);
+        GuestAgent::append_submsg_type(&mut buf, MsgSubType::SubMsgEnd);
+        self.stream.write_all(&buf)?;
+        self.get_response().unwrap();
+        Ok(())
+    }
+
+    fn run_process(
+        &mut self,
+        bin: &[u8],
+        argv: &[&[u8]],
+        maybe_env: Option<&[&[u8]]>,
+        uid: u32,
+    ) -> Result<(), io::Error> {
+        let mut buf = GuestAgent::create_header(self.get_new_msg_id(), MsgType::MsgRunProcess);
+        GuestAgent::append_submsg_bytes(&mut buf, MsgSubType::SubMsgRunProcessBin, bin);
+        GuestAgent::append_submsg_array(&mut buf, MsgSubType::SubMsgRunProcessArg, argv);
+        match maybe_env {
+            Some(env) => {
+                GuestAgent::append_submsg_array(&mut buf, MsgSubType::SubMsgRunProcessEnv, env)
+            }
+            None => (),
+        }
+        GuestAgent::append_submsg_u32(&mut buf, MsgSubType::SubMsgRunProcessUid, uid);
+        GuestAgent::append_submsg_type(&mut buf, MsgSubType::SubMsgEnd);
+        self.stream.write_all(&buf)?;
+        self.get_response().unwrap();
+        Ok(())
+    }
+}
+
 fn main() {
-    println!("Hello, world!");
+    let mut child = Command::new("qemu-system-x86_64")
+        .args(&[
+            "-m", "256m",
+            "-nographic",
+            "-vga", "none",
+            "-kernel", "init-container/vmlinuz-virt",
+            "-initrd", "init-container/initramfs.cpio.gz",
+            "-no-reboot",
+            "-net", "none",
+            "-smp", "1",
+            "-append", "console=ttyS0 panic=1",
+            "-device", "virtio-serial",
+            "-chardev", "socket,path=./manager.sock,server,nowait,id=manager_cdev",
+            "-device", "virtserialport,chardev=manager_cdev,name=manager_port",
+            "-drive", "file=./squashfs_drive,cache=none,readonly=on,format=raw,if=virtio"])
+        .stdin(Stdio::null())
+        .spawn()
+        .expect("failed to spawn VM");
+
+    let mut ga = GuestAgent::connected("./manager.sock").unwrap();
```

Comment:
Is it possible to wait indefinitely?

Reply:
I guess there could be some arbitrary timeout / number of tries we could pick. Is that what you're suggesting?

Reply:
Yes, if it's needed. I was wondering whether it is possible that the socket is never created / no longer exists when the process is properly spawned

Reply:
Indeed it was possible, e.g. when process did not have write access to the socket path.